### PR TITLE
Iterators

### PR DESCRIPTION
**Pull request by adsharma**
Just so that the code lives up to the name of the library.

We still need to write up some documentation describing the internals. But at a high level, the iterator iterates over Item -> Item unlike rocksdb::Iterator which is a string -> string map.

This pull request contains only the LimitIterator.

### CODE REVIEW

File: include/iterlib/Iterator.h [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+  LITERAL,
+  FUTURE,
+  ORDERBY,
+  BINARY,
+};
+
+enum class ResultOrder {
+  DEFAULT = 0, // Index order
+  INDEX_ORDER = 0,
+  ID_ORDER,
```
Comment (manukranthk) [inline include/iterlib/Iterator.h]: We can drop specific order types here.

File: include/iterlib/Item.h [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+      syncMap();
+    }
+  }
+
+  const dynamic& value() const {
+    return *this;
+  }
+
+  bool operator<(const ItemOptimized& other) const {
+    return std::tie(value(), ts_, id_)
```
Comment (yfeldblum) [inline include/iterlib/Item.h]: Provide a `private` member function (e.g., `tied`). Then this function will look like:

```
return tied() < other.tied();
```

DRYer, less risk of ordering bugs.

File: include/iterlib/Item.h [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+// =  => equal
+// -1 => not equal or less than. either uncomparable or greater than
+// Notice that this signature is different from std:: compare (two value)
+// and also different from C-style compare (greater/equal/less) which assumes
+// total ordering. This function signature is usable by comparators, and also
+// the upcoming join operator (which uses useTsIdBackup = false)
+//
+// The function performs a projection as well as comparison and deals with
+// corner cases involving some columns not being present in v1 or v2
+int lessThanOrEqual(const Item& v1,
```
Comment (yfeldblum) [inline include/iterlib/Item.h]: Indentation. Formal parameters start on the next line, are indented +4.

File: include/iterlib/Item.h [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+            const std::vector<std::string>& columns,
+            const std::vector<bool>& isColumnDescending,
+            bool useTsIdBackup);
+
+int lessThanOrEqual(const Item& v1,
+            const Item& v2,
+            const std::vector<std::string>& columns,
+            bool useTsIdBackup);
+
+inline std::ostream& operator<<(std::ostream& os, const Item& row) {
```
Comment (yfeldblum) [inline include/iterlib/Item.h]: Make this non-inline and move this out of the header. Then the header will need to include only `<iosfwd>`, not `<ostream>` or `<iostream>`.

File: include/iterlib/LimitIterator.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  std::string cookie() const override {
+    return innerIter_->cookie();
+  }
+
+  virtual bool orderPreserving() const override {
+    return true;
+  }
+
+protected:
+ bool doNext() override;
```
Comment (yfeldblum) [inline include/iterlib/LimitIterator.h]: Indentation.

File: src/Item.cpp [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    } else {
+      const auto& attr1 = m1.atNoThrow(attrName);
+      const auto& attr2 = m2.atNoThrow(attrName);
+
+      try {
+        attrEqual = (attr1 == attr2);
+        attrLess = attrEqual ? false : (isDescending ? (attr1 < attr2)
+                                                     : (attr1 > attr2));
+      } catch (const std::exception& ex) {
+        // == does not throw, but < does, protecting against such cases
```
Comment (yfeldblum) [inline src/Item.cpp]: wut

File: src/Item.cpp [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  if (useTsIdBackup) {
+    if (v1.ts() != v2.ts()) {
+      return v1.ts() < v2.ts();
+    }
+
+    if (v1.id() != v2.id()) {
+      return v1.id() < v2.id();
+    }
```
Comment (yfeldblum) [inline src/Item.cpp]: Use `std::tie`.

File: src/Item.cpp [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  return -1;
+}
+
+int lessThanOrEqual(const Item& v1,
```
Comment (yfeldblum) [inline src/Item.cpp]: Parameter indentation.

File: src/WrappedIterator.cpp [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+namespace iterlib {
+
+folly::Future<folly::Unit> WrappedIterator::prepare() {
+  if (prepared_) {
+    return folly::makeFuture();
+  }
+  if (innerIter_) {
+    return innerIter_->prepare()
+      .then([this]() {
+          // Code to extract any data from inner iter
```
Comment (yfeldblum) [inline src/WrappedIterator.cpp]: Indentation. Lambda code block is indented +2 compared to outer scope (this is +4).

File: src/WrappedIterator.cpp [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+          // Hopefully none in a zero-copy implementation
+        })
```
Comment (yfeldblum) [inline src/WrappedIterator.cpp]: Closing `})` should not be indented compared to the opening `{(`.

File: tests/IteratorTest.h [hunk 11]
```diff
@@ -0,0 +1,27 @@
+#pragma once
```
Comment (yfeldblum) [inline tests/IteratorTest.h]: Rename this header....

File: include/iterlib/Item.h [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+  using dynamic::getRef;
+  using dynamic::getNonConstRef;
+
+  static const Item kEmptyItem;
+  Item() : dynamic() {}
+  Item(const dynamic& other) : dynamic(other) {}
+
+  Item(const Item& other) = default;
+  Item& operator=(const Item& other) = default;
+  Item& operator=(Item&& other) = default;
```
Comment (adsharma) [inline include/iterlib/Item.h]: will do.

File: include/iterlib/Item.h [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+  std::string toString() const {
+    return folly::sformat("[{0}, {1}, {2}]", ts(), id(), dynamic::toString());
+  }
+
+protected:
+  id_t id_;
+  int64_t ts_;
+};
+
+// lessThanOrEqual is a 3-value function:
```
Comment (adsharma) [inline include/iterlib/Item.h]: Looks cleaner. Will change.

File: tests/IteratorTest.cpp [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+  const auto res = std::vector<ItemOptimized>{
+    {1, 0, unordered_map_t{{"int1", 2L}, {"str1", std::string("banana")},
+                  {"int2", 3L}}},
+    {2, 0, unordered_map_t{{"int1", 1L}, {"str1", std::string("orange")},
+                 {"int2", 3L}}},
+    {3, 0, unordered_map_t{{"int1", 1L}, {"str1", std::string("apple")},
+                 {"int2", 2L}}},
+    {4, 0, unordered_map_t{{"int1", 1L}, {"str1", std::string("banana")},
+                 {"int2", 4L}}},
+  };
```
Comment (adsharma) [inline tests/IteratorTest.cpp]: @manukranthk  suggests clang-format -i on all the code.

File: tests/IteratorTest.h (hunk 11 above)
Comment (adsharma) [inline tests/IteratorTest.h]: What's wrong with IteratorTest.h? Did you mean something like ExpectIterator.h?

File: include/iterlib/Item.h (hunk 13 above)
Comment (manukranthk) [inline include/iterlib/Item.h]: We can call the it 
Ordering => predicate?

File: tests/IteratorTest.h [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+
+#include <gtest/gtest.h>
+#include <folly/io/async/EventBaseManager.h>
+
+#include "iterlib/Iterator.h"
+
+namespace iterlib {
+
+template <typename T>
+void ExpectIterator(Iterator* it, const std::vector<T>& expected) {
```
Comment (adsharma) [inline tests/IteratorTest.h]: https://github.com/google/googletest/blob/master/googlemock/docs/CookBook.md

has some documentation on `testing::MatcherInterface`. Will read up.

**Comment by adsharma** [comment]
Ran clang-format on all the changed files. Let me know if you still see any style/indentation issues.

File: tests/IteratorTest.cpp [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+  auto it = folly::make_unique<FutureIterator<ItemOptimized>>(
+    folly::makeFuture(res));
+  auto limitIt = folly::make_unique<LimitIterator>(it.release(), 2, 0);
+  const auto expected = std::vector<Item>{res[0], res[1]};
+  ExpectIterator(limitIt.get(), expected);
+}
+
+int main(int argc, char* argv[]) {
```
Comment (adsharma) [inline tests/IteratorTest.cpp]: For people installing folly via rpms/debs, TestMain.cpp is not a part of libfolly.so.
It's not clear how to link to it.

**Comment by adsharma** [comment]
Update addresses many review comments so far. Still TODO - custom gtest matcher for ExpectIterator.

File: tests/IteratorTest.cpp (hunk 16 above)
Comment (yfeldblum) [inline tests/IteratorTest.cpp]: See `folly/test/Makefile.am`:

https://github.com/facebook/folly/blob/v2016.11.14.00/folly/test/Makefile.am#L19-L34

File: include/iterlib/Item.h [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+ protected:
+  id_t id_;
+  int64_t ts_;
+};
+
+// partialCompare is a 3-value function:
+//
+// LT  => less than
+// EQ  => equal
+// GT_OR_NOT_COMPARABLE  => uncomparable or greater than
```
Comment (yfeldblum) [inline include/iterlib/Item.h]: Why are `GT` and `NOT_COMPARABLE` put together?

File: include/iterlib/Item.h [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+
+  Item(const Item& other) = default;
+  Item& operator=(const Item& other) = default;
+  Item& operator=(Item&& other) = default;
+
+  bool operator==(const Item& other) const {
+    return dynamic::operator==(static_cast<const dynamic&>(other));
+  }
+
+  id_t id() const {
```
Comment (adsharma) [inline include/iterlib/Item.h]: Docs in Internals.md

File: tests/IteratorTest.cpp (hunk 16 above)
Comment (adsharma) [inline tests/IteratorTest.cpp]: make install doesn't seem to install TestMain.cpp in a library of its own. The method above works only for tests in folly itself.

Would like to keep it simple for target users of iterlib for now.

File: include/iterlib/Item.h (hunk 17 above)
Comment (adsharma) [inline include/iterlib/Item.h]: Because I wasn't brave enough to make the changes in one step. Will break it into two.

**Comment by adsharma** [comment]
I think the updated diff addresses most of the comments. Please let me know if I missed anything.

https://paste.ubuntu.com/23497246/

Is how far I got on using EXPECT_THAT. It's blocked by the gtest on my box not being new enough to support MatchResultListener. Ok to defer to another diff?

File: tests/IteratorTest.cpp (hunk 16 above)
Comment (yfeldblum) [inline tests/IteratorTest.cpp]: Ah I guess that is something to be fixed in Folly.

**Comment by yfeldblum** [comment]
Okay to defer using gmock to another PR.

**Comment by yfeldblum** [comment]
And okay to defer making `partialCompare` do a real partial compare (4-valued).

**Comment by adsharma** [comment]
partialCompare already returns 4 values. Opened #8 for ExpectIterator

File: include/iterlib/Iterators.h [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+ * Please try to include only iterators you need.
+ */
+
+#include "iterlib/Iterator.h"
+
+// Leaf Iterators
+#include "iterlib/RocksDBIterator.h"
+
+// Most common Iterators
+#include "iterlib/FilterIterator.h"
```
Comment (manukranthk) [inline include/iterlib/Iterators.h]: These iterators aren't added yet.

File: include/iterlib/LimitIterator.h [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+//  This source code is licensed under the BSD-style license found in the
+//  LICENSE file in the root directory of this source tree. An additional grant
+//  of patent rights can be found in the PATENTS file in the same directory.
+#pragma once
+
+#include "iterlib/WrappedIterator.h"
+
+namespace iterlib {
+
+class LimitIterator : public WrappedIterator {
```
Comment (manukranthk) [inline include/iterlib/LimitIterator.h]: Should we add block comments for the classes to explain the behavior of the specific iterators?

File: src/Item.cpp [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+const std::string kTimeKey(QueryConstants::kTimeKey);
+
+namespace {
+const dynamic kEmptyItemInternal{variant::unordered_map_t{
+    {{":id", int64_t(std::numeric_limits<id_t>::max())}, {":time", 0}}}};
+}
+
+const Item Item::kEmptyItem{kEmptyItemInternal};
+
+PartialOrder partialCompare(const Item& v1, const Item& v2,
```
Comment (manukranthk) [inline src/Item.cpp]: Might be good to add some documentation explaining a parallel to SQL
`ORDER BY column_name ASC|DESC, column_name ASC|DESC; `

File: include/iterlib/Iterator.h [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+  }
+
+  bool isDone_; // a flag to check if the iterator is done
+  Item key_;
+
+  bool advancedAtleastOnce_;
+
+  // whether this iterator has been prepared. Only after the iterator has
+  // been prepared, we can call next() and access all the getters.
+  bool prepared_;
```
Comment (manukranthk) [inline include/iterlib/Iterator.h]: Should this be std::atomic ?

File: include/iterlib/LimitIterator.h (hunk 20 above)
Comment (manukranthk) [inline include/iterlib/LimitIterator.h]: Especially things like
startOffset_

File: include/iterlib/Iterators.h (hunk 19 above)
Comment (adsharma) [inline include/iterlib/Iterators.h]: Will drop this file.

File: include/iterlib/QueryConstants.h [hunk 23]
```diff
@@ -0,0 +1,17 @@
+//  Copyright (c) 2016, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under the BSD-style license found in the
+//  LICENSE file in the root directory of this source tree. An additional grant
+//  of patent rights can be found in the PATENTS file in the same directory.
+
+#pragma once
+
+namespace iterlib {
+namespace QueryConstants {
```
Comment (adsharma) [inline include/iterlib/QueryConstants.h]: Will remove

File: include/iterlib/LimitIterator.h (hunk 20 above)
Comment (adsharma) [inline include/iterlib/LimitIterator.h]: I have a spec which defines all these behaviors in more detail. Will work on getting that documentation polished before adding to the repo. I can add a one line description here though.

File: src/Item.cpp (hunk 21 above)
Comment (adsharma) [inline src/Item.cpp]: I couldn't come up with a SQL analog (only one row here). Added a simpler comment.

File: include/iterlib/Iterator.h (hunk 22 above)
Comment (adsharma) [inline include/iterlib/Iterator.h]: We'll need a lock to protect the state being prepared anyway.

**Comment by adsharma** [comment]
Could you take a look at the updated diff?

File: include/iterlib/FutureIterator.h [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+
+/**
+ * Provides Iterator interface for arbitrary future which returns
+ * std::vector<Item> result.
+ * Particularly useful for data source APIs which do not have iterator interface
+ * and in tests.
+ */
+template <typename T = Item> class FutureIterator : public Iterator {
+ public:
+  explicit FutureIterator(folly::Future<std::vector<T>>&& f,
```
Comment (manukranthk) [inline include/iterlib/FutureIterator.h]: We can generalize the FutureIterator to have a lazy yield style pagination/next methods.

File: include/iterlib/Item.h [hunk 25]
```diff
@@ -0,0 +1,202 @@
...
+    } catch (const std::exception& ex) {
+      LOG_EVERY_N(WARNING, 1000) << ex.what();
+    }
+  }
+
+  void sync() {
+    // This is a heuristic. The actual sync depends on what was
+    // modified. Use syncIdTs/syncMap directly depending on the
+    // context.
+    if (id_ == 0) {
```
Comment (manukranthk) [inline include/iterlib/Item.h]: This should be a self documenting constant.

File: include/iterlib/Iterator.h [hunk 26]
```diff
@@ -0,0 +1,211 @@
...
+  explicit Iterator(IteratorType type = IteratorType::NONE);
+  Iterator(const Iterator&) = delete;
+  Iterator& operator=(const Iterator&) = delete;
+
+  Iterator(Iterator&&) = default;
+  Iterator& operator=(Iterator&&) = default;
+
+  virtual ~Iterator() {}
+
+  static id_t max() { return std::numeric_limits<id_t>::max(); }
```
Comment (manukranthk) [inline include/iterlib/Iterator.h]: Are min max functions required in Iterator?

File: include/iterlib/FutureIterator.h (hunk 24 above)
Comment (adsharma) [inline include/iterlib/FutureIterator.h]: The rocksdb iterator in #9 is such a lazy iterator. Sure, we can look into some lazy iterators for testing purposes (or just use rocksdb iterator for now).

File: include/iterlib/Iterator.h (hunk 26 above)
Comment (adsharma) [inline include/iterlib/Iterator.h]: Not in this pull request, but some of the upcoming pull requests use them.

**Pull request merged**